### PR TITLE
 Support custom domain verification

### PR DESCRIPTION
**Pull request by MarkRunWu**
#7

### CODE REVIEW

File: internal/cron/verify_domain_ownership.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+			domainVerifications = append(domainVerifications, verifiedRecord)
+		}
+		for _, unverifiedRecord := range unverifedRecords {
+			domainVerifications = append(domainVerifications, unverifiedRecord)
+		}
+
+		var consumedCount = 0
+		for _, domainVerification := range domainVerifications {
+			key, value := domainVerification.GetTxtRecord()
+			values, err := net.LookupTXT(key)
```
Comment (MarkRunWu) [inline internal/cron/verify_domain_ownership.go:44]: @kiootic plan to add test case for this worker. May i know if you have any prefer way other than patched with https://pkg.go.dev/github.com/foxcpp/go-mockdns#section-readme to mock out this line for testing ?

File: internal/handler/controller/config.go [hunk 2]
```diff
@@ -6,14 +6,15 @@ import (
...
-	ACL               *watch.File[config.ACL]
+	MaxDeploymentSize  int64
+	StorageKeyPrefix   string
+	HostIDScheme       config.HostIDScheme
+	HostPattern        *config.HostPattern
+	ReservedApps       map[string]struct{}
+	TokenAuthority     string
+	TokenSigningKey    []byte
+	ACL                *watch.File[config.ACL]
+	DomainVerification bool
```
Comment (kiootic) [inline internal/handler/controller/config.go:17]: `DomainVerificationEnabled`

File: internal/config/app_domain.go [hunk 3]
```diff
@@ -1,6 +1,6 @@
 package config
 
 type AppDomainConfig struct {
-	Domain string `json:"domain" pageship:"required,max=200,hostname_rfc1123,lowercase"`
+	Domain string `json:"domain" pageship:"required,max=200,hostname_port|hostname_rfc1123,lowercase"`
```
Comment (kiootic) [inline internal/config/app_domain.go:4]: I don't think we need port? Domain verification/association should not care about the port.

File: internal/handler/controller/domain.go [hunk 4]
```diff
@@ -5,40 +5,84 @@ import (
...
 
-func (c *Controller) makeAPIDomain(domain *models.Domain) *apiDomain {
-	return &apiDomain{Domain: domain}
+type apiDomainVerification struct {
+	*models.DomainVerification
+}
+
+func (c *Controller) makeAPIDomainVerification(domainVerification *models.DomainVerification) *apiDomainVerification {
+	return &apiDomainVerification{DomainVerification: domainVerification}
+}
```
Comment (kiootic) [inline internal/handler/controller/domain.go:24]: Unused?

File: internal/handler/controller/domain.go [hunk 5]
```diff
@@ -50,6 +94,9 @@ func (c *Controller) handleDomainCreate(w http.ResponseWriter, r *http.Request)
 		if !ok {
 			return nil, models.ErrUndefinedDomain
 		}
+		if c.Config.DomainVerification {
+			return nil, models.ErrDomainVerificationNotSupported
+		}
```
Comment (kiootic) [inline internal/handler/controller/domain.go:99]: If domain verification is disabled, it means verification is not required for custom domains.

File: internal/handler/controller/domain.go [hunk 6]
```diff
@@ -5,40 +5,84 @@ import (
...
 
+func (c *Controller) handleDomainActivate(w http.ResponseWriter, r *http.Request) {
+	app := get[*models.App](r)
+	domainName := chi.URLParam(r, "domain-name")
+
+	config, ok := app.Config.ResolveDomain(domainName)
+	if !ok {
+		respond(w, func() (any, error) { return nil, models.ErrUndefinedDomain })
+	}
+	if c.Config.DomainVerification {
```
Comment (kiootic) [inline internal/handler/controller/domain.go:64]: I would not expect the activate API would also create the domain; the API flow should be like this:
1. User update app config in CLI
2. Call update config API
3. User activates domain in CLI
4. Call create domain API
5. If domain verification is not enabled: API create domain association.
6. If domain verification is enabled: API create domain verification request.
7. User see the TXT record in CLI
8. User update DNS with the TXT record
9. User activates domain in CLI
10. Call create domain API
11. Domain verification request present: mark the domain verification request as pending.

File: internal/api/models.go [hunk 7]
```diff
@@ -27,7 +27,8 @@ type APIDeployment struct {
 }
 
 type APIDomain struct {
-	*models.Domain
+	Domain             *models.Domain
+	DomainVerification *models.DomainVerification
```
Comment (kiootic) [inline internal/api/models.go:31]: Please gives explicit `json` tags to have camelCase field names.

File: internal/cron/verify_domain_ownership.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	MaxConsumeActiveDomainCount  uint
+	MaxConsumePendingDomainCount uint
+}
+
+func (v *VerifyDomainOwnership) Name() string { return "verify-domain-ownership" }
+
+func (v *VerifyDomainOwnership) CronSchedule() string { return v.Schedule }
+
+func (v *VerifyDomainOwnership) Run(ctx context.Context, logger *zap.Logger) error {
+	now := time.SystemClock.Now().UTC()
```
Comment (kiootic) [inline internal/cron/verify_domain_ownership.go:26]: Pass the clock in struct field.

File: internal/cron/verify_domain_ownership.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+			}
+			consumedCount += 1
+			matched := false
+			for _, v := range values {
+				if v == value {
+					matched = true
+					break
+				}
+			}
+			domain, err := c.GetDomainByName(ctx, domainVerification.Domain)
```
Comment (kiootic) [inline internal/cron/verify_domain_ownership.go:57]: The error is not checked correctly: if `matched` is true and `err` is non-nil (and not `ErrDomainNotFound`), it is ignored.

Please move the error checking right after the operation.

File: internal/cron/verify_domain_ownership.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+func (v *VerifyDomainOwnership) Run(ctx context.Context, logger *zap.Logger) error {
+	now := time.SystemClock.Now().UTC()
+
+	return db.WithTx(ctx, v.DB, func(c db.Tx) error {
+		verified := true
+		unverified := false
+		domainVerifications := []*models.DomainVerification{}
+		verifiedRecords, _ := c.ListDomainVerifications(ctx, nil, &v.MaxConsumeActiveDomainCount, &verified)
+		unverifedRecords, _ := c.ListDomainVerifications(ctx, nil, &v.MaxConsumePendingDomainCount, &unverified)
```
Comment (kiootic) [inline internal/cron/verify_domain_ownership.go:33]: I'd prefer to have an explicit method for this purpose, something like `ListLeastRecentlyCheckedDomain(isVerified, count)`.
- The order of the listing is not (?) part of the API interface. For example, one may found ordering by domain name is more natural.
- Omitting app ID should not be possible for normal list.
- The currently listing method is too complex with many string joins.

File: internal/cron/verify_domain_ownership.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+						))
+					}
+				}
+			} else {
+				domainVerification.VerifiedAt = nil
+				if domain != nil {
+					c.DeleteDomain(ctx, domain.ID, now)
+				}
+			}
+			domainVerification.UpdatedAt = now
```
Comment (kiootic) [inline internal/cron/verify_domain_ownership.go:79]: I'm feeling reserved on using `UpdatedAt` as trigger for verification; updates could be caused by operations other than domain verification requests.

Also seems that if verification failed, the flow is stuck unless user delete the domain and re-add it.

I would suggest to add two columns for this: `last_checked_at` and `will_check_at`.
- The cron job lookup N expired verification requests (i.e. `will_check_at` is before now)
- Update `last_checked to be now after checking.
- If check passed: update `will_check_at` to `now + X` (X could be arbitrary value, like 1h).
- If check failed: update `will_check_at` to be null to stop cron job from checking it.
- If user requested check, set `will_check_at` to now.
- Display `last_checked_at` to user to show that the server is indeed running the check process.

File: .github/workflows/test.yaml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        run: |
+          mkdir -p data.local/
+          mkdir -p data.local/storage
+          echo "TEST_PAGESHIP_DATABASE_URL=sqlite:///${PWD}/data.local/test_data.db" >> "$GITHUB_ENV"
+          echo "TEST_PAGESHIP_STORAGE_URL=file:///${PWD}/data.local/storage" >> "$GITHUB_ENV"
+      - name: Test 
+        run: go test ./...
+        env:
+          TEST_PAGESHIP_HOST_PATTERN: http://*.localtest.me
+          TEST_PAGESHIP_STORAGE_KEY_PREFIX: pageship/
```
Comment (kiootic) [inline .github/workflows/test.yaml:30]: These should be setup in the `testutil` package. I would expect I can just run `go test ./...` and all tests should be run without further config. The database/storage directory can be a temp dir.

File: internal/cron/verify_domain_ownership.go (hunk 1 above)
Comment (kiootic) [reply internal/cron/verify_domain_ownership.go:44]: I would suggest making a interface for this, just like the `Clock` interface.

```go
type DNSLookuper interface {
  LookupTXT(name string) ([]string, error)
}


var SystemDNS DNSLookuper = systemDNS{}

type systemDNS struct {}

func (s systemDNS) LookupTXT(name string) ([]string, error) {
  return net.LookupTXT(name)
}


type MockDNS struct {
    TXTRecords map[string][]string
}

func (s *MockDNS) LookupTXT(name string) ([]string, error) {
  /* ... */
}

File: internal/cron/verify_domain_ownership.go (hunk 11 above)
Comment (MarkRunWu) [reply internal/cron/verify_domain_ownership.go:79]: OH, i thought it validate every minutes or we can put X in app config then only `last_checked_at` is required as `will_check_at` is derived value.

Comment (kiootic) [reply internal/cron/verify_domain_ownership.go]: Not sure how would it handle the 4th/5th point. If just `last_checked_at` is used, how to distinguish between a failed check and pending check, and allow user to retry the check?

Comment (MarkRunWu) [reply internal/cron/verify_domain_ownership.go]: You are right. The two fields are essentials for saving worker from busy.

**Comment by MarkRunWu** [comment]
@kiootic Updated, 🙇 

Here tui changes:

`pageship domains --app=<app_id>`

截圖 2023-12-14 下午4 48 51


`pageship domains --app=<app_id> activate <domain>`

截圖 2023-12-14 下午4 49 13

File: internal/db/sqlite/domain_verification.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+	err := sqlx.SelectContext(ctx, q.ext, &domainVerifications, stmt, time, count)
+	if err != nil {
+		return nil, err
+	}
+	return domainVerifications, err
+}
+
+func (q query[T]) ScheduleDomainVerificationAt(ctx context.Context, id string, time time.Time) error {
+	_, err := q.ext.ExecContext(ctx, `
+    UPDATE domain_verification SET
```
Comment (kiootic) [inline internal/db/sqlite/domain_verification.go:132]: Need updated at too

File: internal/handler/controller/authn.go [hunk 14]
```diff
@@ -139,8 +139,11 @@ func (c *Controller) middlewareAuthn(next http.Handler) http.Handler {
 			loggers.Logger = loggers.Logger.With(zap.String("user", authn.Subject))
 			loggers.authn = loggers.Logger
 
-			entry := middleware.GetLogEntry(r).(*httputil.LogEntry)
-			entry.Logger = entry.Logger.With(zap.String("user", authn.Subject))
+			entry := middleware.GetLogEntry(r)
+			if entry != nil {
```
Comment (kiootic) [inline internal/handler/controller/authn.go:143]: No need check nil?

File: internal/handler/controller/controller.go [hunk 15]
```diff
@@ -93,7 +93,13 @@ func (c *Controller) Handler() http.Handler {
...
 
 					r.With(c.requireAccessAdmin()).Route("/{domain-name}", func(r chi.Router) {
-						r.With(c.requireAccessDeployer()).Post("/", c.handleDomainCreate)
+						r.With(c.requireAccessDeployer()).Post("/", func(w http.ResponseWriter, r *http.Request) {
+							if c.Config.DomainVerificationEnabled {
+								c.handleDomainVerification(w, r)
+							} else {
+								c.handleDomainCreate(w, r)
+							}
+						})
```
Comment (kiootic) [inline internal/handler/controller/controller.go:102]: Can we merge the two functions together?

File: internal/models/domain_verification.go [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+func NewDomainVerification(now time.Time, domain string, appID string) *DomainVerification {
+	num1, err := rand.Int(rand.Reader, big.NewInt(999999))
+	if err != nil {
+		panic(err)
+	}
+	num2, err := rand.Int(rand.Reader, big.NewInt(999999))
+	if err != nil {
+		panic(err)
+	}
```
Comment (kiootic) [inline internal/models/domain_verification.go:32]: You can reference `RandomID` function instead to use random bytes directly.

**Comment by MarkRunWu** [comment]
@kiootic updated 🙇

**Review by kiootic — APPROVED** [review]
Thanks!

**Pull request merged**